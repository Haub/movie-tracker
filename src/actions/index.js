export const displayHanksMovies = (movies) => ({
  type: 'DISPLAY_HANKS_MOVIES',
  movies
})

export const updateFavorites = (movies) => ({
  type: 'UPDATE_FAVORITES',
  movies
export const addFavorite = (movie) => ({
  type: 'ADD_FAVORITE',
  movie
});

export const removeFavorite = (movieId) => ({
  type: 'REMOVE_FAVORITE',
  movieId
});

export const loginUser = (user) => ({
  type: 'LOGIN_USER',
  user
});

export const logoutUser = () => ({
  type: 'LOGOUT_USER'
});
