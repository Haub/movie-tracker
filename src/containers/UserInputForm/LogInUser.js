import React, {Component} from 'react';
import { NavLink, withRouter } from 'react-router-dom';
import {connect} from 'react-redux';

import './Navigation.css';
import { userLogIn, fetchUserFavorites } from '../../Utilities/fetchApi';
import { loginUser, updateFavorites, alertUser } from '../../actions';
import Alert from '../Alert';
 
export class LoginUser extends Component {
  constructor() {
    super()
    this.state = {
      email: '',
      password: ''
    }
  }

  handleChange = (e) => {
    const { value, name } = e.target;
    this.setState({
      [name] : value
    });
  }

  handleSubmit = async (e) => {
    e.preventDefault();
    const { login, history, updateFavorites, alertUser } = this.props;
    const userInfo = await userLogIn(this.state);
    if (userInfo.alert) {
      return alertUser(userInfo.alert)
    }
    const userFavorites = await fetchUserFavorites(userInfo.data.id);
    updateFavorites(userFavorites.data);
    login(userInfo.data);
    history.push('/user');
  }

  render() {
    const { email, password } = this.state;
    const { alertUser } = this.props;
    return (
      <div className="NavBar">
        <form onSubmit={this.handleSubmit}>
          <input 
            required
            type='email'
            placeholder='Email'
            value={email}
            name='email'
            onChange={this.handleChange}
          />
          <input
            required
            placeholder='Password'
            value={password}
            name='password'
            onChange={this.handleChange}
          />
          <button>Login</button>
          <NavLink className="NavLink" to='/signup' onClick={() => alertUser('')}>
            Sign Up
          </NavLink>
        </form>
        <Alert />
      </div>
    )
  }
}

const mapStateToProps = (state) => ({ 
  currentUser: state.user 
})

const mapDispatchToProps = (dispatch) => ({
  login: (user) => dispatch(loginUser(user)),
  updateFavorites: (movies) => dispatch(updateFavorites(movies)),
  alertUser: (message) => dispatch(alertUser(message))
})

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(LoginUser));
