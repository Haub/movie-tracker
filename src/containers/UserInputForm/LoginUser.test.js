import React from 'react';
import { shallow } from 'enzyme';

import { LogInUser, mapDispatchToProps } from './LogInUser';
import { loginUser, updateFavorites, alertUser } from '../../actions';


describe('LogInUser', () => {
  describe('LogInUser Component', () => {
    let wrapper;
    let login; 
    let history; 
    let updateFavorites; 
    let alertUser;
    let e;

    beforeEach(() => {
      login = jest.fn(); 
      history = { push: jest.fn().mockImplementation(() => ({location:'/user'}))}
      updateFavorites = jest.fn();
      alertUser = jest.fn();
      e = {preventDefault:jest.fn()}

      window.fetch = jest.fn().mockImplementation(() => (
        Promise.resolve({ json:() => Promise.resolve({data:{id:1}})})
      ))

      wrapper = shallow(
        <LogInUser 
          login={login}
          history={history}
          updateFavorites={updateFavorites}
          alertUser={alertUser}
        />)
    })

    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should call a reset on alert user on link to sign up', () => {
      wrapper.find('NavLink').simulate('click');
      expect(alertUser).toHaveBeenCalled();
    });

    describe('handleChange', () => {
      it('should set state when invoked', () => {
        const wrapper = shallow(<LogInUser />);
        let e = {target:{value:'foo', name:'email'}}

        wrapper.instance().handleChange(e);
        expect(wrapper.state().email).toEqual('foo');

        e = {target:{value:'foo', name:'password'}}
        wrapper.instance().handleChange(e);
        expect(wrapper.state().password).toEqual('foo');
      });
    });

    describe('handleSubmit', () => {
      beforeEach(() => {
        window.fetch = jest.fn().mockImplementation(() => (
          Promise.resolve({ json:() => Promise.resolve({data:{id:1}})})
        ))
        login = jest.fn(); 
        history = { push: jest.fn().mockImplementation(() => ({location:'/user'}))}
        updateFavorites = jest.fn();
        alertUser = jest.fn();
        e = {preventDefault:jest.fn()}

        wrapper = shallow(
          <LogInUser 
            login={login}
            history={history}
            updateFavorites={updateFavorites}
            alertUser={alertUser}
          />)
      })

      it('should call alertUser if fetch rejected', async () => {
        window.fetch = jest.fn().mockImplementation(() => {
          Promise.resolve({ json:() => Promise.reject()})
        })
        await wrapper.instance().handleSubmit(e);
        expect(alertUser).toHaveBeenCalled();
      });
  
      it('should fetch user favorites if log in fetch passes', async () =>{
        await wrapper.instance().handleSubmit(e);
        expect(updateFavorites).toHaveBeenCalled();
    });

      it('should log in user if fetch passes', async () =>{
        await wrapper.instance().handleSubmit(e);
        expect(login).toHaveBeenCalled();
      });
    
  });

  describe('mapDispatchToProps', () => {

  });
});