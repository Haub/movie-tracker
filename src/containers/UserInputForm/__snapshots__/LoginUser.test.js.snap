// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LogInUser LogInUser Component should match snapshot 1`] = `
<div
  className="NavBar"
>
  <form
    onSubmit={[Function]}
  >
    <input
      className="email-input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      required={true}
      type="email"
      value=""
    />
    <input
      className="password-input"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      required={true}
      value=""
    />
    <button
      className="login-btn"
    >
      Login
    </button>
  </form>
  <NavLink
    activeClassName="active"
    aria-current="page"
    className="login-link"
    onClick={[Function]}
    to="/signup"
  >
    Sign Up
  </NavLink>
  <Connect(Component) />
</div>
`;
